### PR TITLE
T44: Create message modal

### LINKED ISSUE #44: New message modal
Upon clicking a "new message" button, a modal should pop up that allows users to search up other users who they can start a new conversation with

The implementation should follow a similar pattern to #36

### PR DESCRIPTION
**Pull request by NasihNazeem**
This feature allows users to create a new message to another user through a CreateMessage button on the `Messages.tsx` page.

This modal provides a list of users the active user may have recently messaged for quick action, and a search bar for the active user to search for any users they currently follow. The constraint with this search bar is that it only displays people that the active user has followed.

Once you click on the search icon, the modal routes you to a messages page where you can begin your conversation with the selected user. The actual messaging functionality is out of scope of this PR.

**TECH DEBT:**
- Currently, the modal has its own `List` and `ListItem` components, those should be updated to create something more universal where we can input JSX as a prop from a parent component.
 
- In the search bar the `Autocomplete` component has a custom `renderOption` attribute where we render a `ListItemText` element that requires a `primary` and `secondary` attribute. These take in elements themselves, but this should be updated so that we can pass this as a prop from a parent component as well.

- Figure out how to change the layout of the dropdown list (the `Autocomplete` options) so that it fills the width of the modal and doesn't look like a "dropdown". Ideally the options should look exactly like the "recent messages" list under the search bar

### CODE REVIEW

**Review by AnindaS13 — CHANGES_REQUESTED** [review]
image

I feel like theres a bit too much whitespace between the magnifying glass and the text beside it. What do you think of moving the text in the search bar over to align with New Message vertically? That might reduce enough whitespace. Or do you think that makes it look weirder. Let me know.

**Comment by AnindaS13** [comment]
d

**Comment by AnindaS13** [comment]
image

I feel like it kinda looks weird with both the displayname and the @username beside each other like that when you select the result in the search bar. The dropdown of peoples names is fine (and probably needed to differentiate two ppl with the same display name). Do you guys think that just having the username in the search tab once you select it would look a little less awkward?

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Nitpick, can this be called open?

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Can this say "Search followers" or something similar? Since we're restricting it to that condition I think that makes more sense

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Maybe it won't matter since we might get rid of this state, but it kinda bothers me how this looks. I think we can move the ternary down into the child component rather than putting it here:

```suggestion
        setSelectedUser={(state) => setSelectedUser(state)}
```

And then in the child component do something like this:

```
// SearchBarMessages.tsx
onChange={(_, value) =>
 value ? setSelectedUser(value) : setSelectedUser({ displayName: "" ... })
```

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: I think there's some overlap here with what I have in #54. I made a new Redux slice with a similar state, so I'd say let's wait until that PR gets merged before we continue with this one cause we might be able to share that store

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: This button seems a little redundant to me in practice. Since we can't add multiple people into a group chat like other apps can, it makes me question if we even need it

I also don't like how we need to click this button to proceed when using the search bar, but we don't need to click it when clicking on the quick select list below. The pattern is just inconsistent. I'd suggest we either ditch the button all together or make the user click this button to proceed in both cases

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Similar story here as with the other state setter, I think I'd prefer if we try to mirror the `onOpen` and `onClose` handlers than do this. So in the child component rather than take in a `setFocusSearchBar` prop, we'd take and `onOpen` and `onClose` prop

```suggestion
        onClose={() => setFocusSearchBar(false)}
        onOpen={() =>setFocusSearchBar(true)}
```

File: src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx]: Can we use a `Box` here to follow our pattern? Let's only use `Stack` if there's a specific reason to

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/SearchBarMessages.tsx]: Can you please remove the extra spaces? Also I think you need the "&" in front of the period right?

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/SearchBarMessages.tsx]: Cool!

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/SearchBarMessages.tsx]: Same here, can we use a `Box` instead?

File: src/pages/Messages.tsx
Comment (Dennull) [inline src/pages/Messages.tsx]: Nitpick, what do you think about changing this to "showMessageModal"? I think it's a bit more descriptive

**Review by Dennull — CHANGES_REQUESTED** [review]
Nice work! I think we should wait until I merge in #54 before we merge this since there's some stuff there that we might be able to use here too

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: I will update this to say "Search following" because it's people you followed rather than the people that follow you

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: updated this, thanks!

For some reason I had it both in the child and parent, but I changed it to your suggestion

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: I can remove the button since we're not going to be doing anything like multiple people in a group

File: src/components/Messages/ConversationList.tsx
Comment (Dennull) [inline src/components/Messages/ConversationList.tsx]: Do we need the index here?

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: That's weird, I thought this was taken care of in `Theme.tsx`. Can you take a look to see why it doesn't work? Our other dialogs have rounded borders without having to set the border radius, so idk why this one isn't the same

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Nit, can use `paddingX` to save some lines

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: I don't see these used right?

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Does this div do anything? If not can we remove it and the corresponding style?

File: src/components/Messages/CreateMessageModal/MessagesModalList.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/MessagesModalList.tsx]: I think I said I'd leave it up to you to decide what to do with this type, did you decide to keep it? Just curious

File: src/components/Messages/CreateMessageModal/MessagesModalList.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/MessagesModalList.tsx]: Do you think we can use the other user's id as the key? You can only have one conversation with each unique user, so I think it makes sense

File: src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx]: Nitpick, this isn't a stack anymore, so can we change the class name?

File: src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx]: Does `primaryTextContainer` do anything? If not can we remove it?

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/SearchBarMessages.tsx:1]: Can we rename this component to `MessagesSearchBar`? I think that sounds better, but let me know what you think. I also proposed some styling changes in #62. Let me know what you think about them!

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (Dennull) [inline src/components/Messages/CreateMessageModal/SearchBarMessages.tsx]: Instead of passing `setFocusSearchBar` as a prop, can we include this state change in the `onClose` function and `onOpen` function and have those as props instead? That way all `onClose` functionality is wrapped up in one function

File: src/pages/Messages.tsx
Comment (Dennull) [inline src/pages/Messages.tsx]: This isn't being used

**Review by Dennull — CHANGES_REQUESTED** [review]
Looks mostly good, just some code clean up and a few small proposals!

File: src/components/Messages/ConversationList.tsx
Comment (NasihNazeem) [reply src/components/Messages/ConversationList.tsx]: Nope, this is outdated code.

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: I think it has to do with me having the set the height of the dialog through `sx` within `PaperProps` attribute. This seems to override the `Theme.tsx` and if I don't maintain a certain height, it just goes back to minHeight

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Thanks, they aren't being used

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: Nope, outdated styling

File: src/components/Messages/CreateMessageModal/MessagesModalList.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/MessagesModalList.tsx]: Yeah I decided to keep it. I felt like it has a different purpose than `SelectedConversation`.

One thing we could do is rename it to something general so that it's used in both components and then apply that type individually for the required state. 

I might do that rather than have duplicate types

File: src/components/Messages/CreateMessageModal/MessagesModalList.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/MessagesModalList.tsx]: Yep this makes sense

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/SearchBarMessages.tsx]: So you want me to send `onOpen` and `onClose` as props to `MessagesSearchBar.tsx`? 

Would look like 
```suggestion
        onOpen={onOpen}
        onClose={onClose}
```

File: src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/MessagesModalListItem.tsx]: Changed it to `container`

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/SearchBarMessages.tsx:1]: Sure thing

File: src/components/Messages/CreateMessageModal/MessagesModalList.tsx
Comment (NasihNazeem) [reply src/components/Messages/CreateMessageModal/MessagesModalList.tsx]: Renamed `SelectedConversation` to `SelectedUser` to fit the criteria for both `OtherUser` and `SelectedCoversation`.

File: src/components/Messages/CreateMessageModal/CreateMessageModal.tsx
Comment (Dennull) [reply src/components/Messages/CreateMessageModal/CreateMessageModal.tsx]: I see, it's fine then

File: src/components/Messages/CreateMessageModal/SearchBarMessages.tsx
Comment (Dennull) [reply src/components/Messages/CreateMessageModal/SearchBarMessages.tsx]: Resolved this offline!

**Review by Dennull — APPROVED** [review]

**Review by AnindaS13 — APPROVED** [review]

**Pull request merged**
